### PR TITLE
Optimize adding datasets to large histories.

### PR DESCRIPTION
**Pull request by jmchilton**

### CODE REVIEW

**Comment by blankenberg** [comment]
Much faster.

```
==> handler2.log <==
galaxy.tools.actions INFO 2016-03-22 15:38:37,666 Handled output named out_file1 for tool Cut1 (162.697 ms)
galaxy.model INFO 2016-03-22 15:38:37,666 Optimizing addition of datasets to history? False
galaxy.model.mapping INFO 2016-03-22 15:38:37,676 history.db_next_hid executed (9.525 ms)
galaxy.model INFO 2016-03-22 15:38:37,677 history.datasets.append executed (0.018 ms)
galaxy.model INFO 2016-03-22 15:38:37,677 history.add_dataset executed (10.379 ms)
galaxy.model INFO 2016-03-22 15:38:37,677 history.add_datasets executed (10.886 ms)
galaxy.tools.actions INFO 2016-03-22 15:38:37,677 Added output datasets to history (11.028 ms)
galaxy.tools.actions INFO 2016-03-22 15:38:37,691 Verified access to datasets for Job[unflushed,tool_id=Cut1] (8.170 ms)
galaxy.tools.actions INFO 2016-03-22 15:38:37,692 Setup for job Job[unflushed,tool_id=Cut1] complete, ready to flush (15.086 ms)
```

**Comment by jmchilton** [comment]
I'm going to just go ahead and take this out of WIP since we have a good set of tests for this kind of thing, I guess this should fix #1831 to some large degree.

Once every job creation is below say 5 seconds - I think the next steps are:
- #1957 
- #1657 
- Create a set of consistent worker threads for job creation throughout a workflow scheduling - reuse the tool burst stuff without the overhead of launching threads.

Ping @blankenberg

**Comment by jmchilton** [comment]
Thanks for pushing this @blankenberg and for the merge - I'm excited to see Galaxy scale to this degree. I'll work on #1957 to get that  `Handled output named out_file1 for tool Cut1 (162.697 ms)` number down.

**Comment by blankenberg** [comment]
Thanks for all your hard work on this @jmchilton !

File: lib/galaxy/model/__init__.py [hunk 1]
```diff
@@ -1173,7 +1173,7 @@ def add_dataset( self, dataset, parent_id=None, genome_build=None, set_hid=True,
         dataset.history = self
         if genome_build not in [None, '?']:
             self.genome_build = genome_build
-        self.datasets.append( dataset )
+        dataset.history_id = self.id
```
Comment (martenson) [inline lib/galaxy/model/__init__.py:1176]: haha, this is a good one :D

**Pull request merged**